Changes in module 'nf-core/krakenuniq/build'
--- modules/nf-core/krakenuniq/build/main.nf
+++ modules/nf-core/krakenuniq/build/main.nf
@@ -8,7 +8,7 @@
         'biocontainers/krakenuniq:1.0.4--pl5321h19e8d03_0' }"
 
     input:
-    tuple val(meta), path(custom_library_dir, stageAs: "library/*"), path(custom_taxonomy_dir, stageAs: "taxonomy"), path(custom_seqid2taxid)
+    tuple val(meta), path(custom_library_dir, stageAs: "library/*"), path(custom_taxonomy_dir, stageAs: "taxonomy/*"), path(custom_seqid2taxid)
 
     output:
     tuple val(meta), path("$prefix/"), emit: db

************************************************************
